File: src/commands/up.ts
import { Command, Options } from "@effect/cli"
import { Console, Effect } from "effect"
import { ConfigService } from "../services/config.js"
import { ProxyService } from "../services/proxy.js"
import { ShellService } from "../services/shell.js"
import { ProjectConfig } from "../schema/config.js"
import { bold, dim, green, red, blue } from "../fmt.js"

const isInside = (cwd: string, path: string) => cwd === path || cwd.startsWith(path + "/")

// ---------------------------------------------------------------------------
// ship up [--open]
// ---------------------------------------------------------------------------

const openOpt = Options.boolean("open")

export const upCommand = Command.make(
  "up",
  { open: openOpt },
  ({ open }) =>
    Effect.gen(function* () {
      const config = yield* ConfigService
      const proxy = yield* ProxyService
      const shell = yield* ShellService

      // Find current workspace from cwd, falling back to a project root checkout
      const workspaces = yield* config.loadWorkspaces()
      const cwd = process.cwd()
      const workspace = workspaces.find((w) => isInside(cwd, w.path))

      let target: { project: string; domain: string; port: number; path: string }
      if (workspace) {
        target = {
          project: workspace.project,
          domain: workspace.proxyDomain,
          port: workspace.port,
          path: workspace.path
        }
      } else {
        const shipConfig = yield* config.loadConfig()
        const entry = Object.entries(shipConfig.projects).find(([, p]) => isInside(cwd, p.path))
        if (!entry) {
          yield* Console.log(`  ${red("✗")} Not inside a workspace or registered project. Use 'ship init' or 'ship create' first.`)
          return
        }
        const [alias, rootProject] = entry
        const domain = rootProject.domain ?? `${alias}.localhost`
        let port = rootProject.port
        if (port === undefined) {
          // Project registered before root routes existed — allocate and persist
          port = yield* proxy.nextPort()
          yield* config.addProject(alias, new ProjectConfig({ ...rootProject, domain, port }))
        }
        target = { project: alias, domain, port, path: rootProject.path }
      }

      const projectConfig = yield* config.getProject(target.project)

      // Ensure proxy is running
      const proxyRunning = yield* proxy.isRunning()
      if (!proxyRunning) {
        yield* proxy.start()
        yield* Console.log(`  ${green("●")} Proxy started.`)
      }

      // Ensure route exists
      yield* proxy.addRoute(target.domain, target.port).pipe(
        Effect.catchAll(() => Effect.void)
      )

      yield* Console.log(`  ${green("●")} ${bold(target.domain)} ${dim("→")} localhost:${blue(String(target.port))}`)

      // Resolve dev command
      const devCmd = projectConfig.commands.dev
      if (!devCmd) {
        yield* Console.log(`  ${dim("No dev command configured. Proxy route is active.")}`)
        return
      }

      const resolvedCmd = devCmd.replace(/\{port\}/g, String(target.port))

      // Open browser after a short delay (in a background fiber)
      if (open) {
        yield* shell.exec("open", [`https://${target.domain}`]).pipe(
          Effect.delay("2 seconds"),
          Effect.catchAll(() => Effect.void),
          Effect.fork
        )
      }

      yield* Console.log(`  ${dim(`Running: ${resolvedCmd}`)}`)
      yield* Console.log("")

      // Run the dev command (blocks until it exits)
      yield* shell.execInDir(target.path, resolvedCmd)
    }).pipe(
      Effect.catchAll((e) =>
        Console.error(`\n  ${red("Error:")} ${e.message}\n`)
      )
    )
)

File: src/commands/up.rs
use crate::domain::workspace_locate::locate_workspace;
use crate::errors::Result;
use crate::fmt::{blue, bold, dim, green, red};
use crate::services::{config, proxy, shell};
use crate::util::cwd_string;

// ---------------------------------------------------------------------------
// ship up [--open]
//
// Locate the current workspace via the pure locator, falling back to a
// registered project root checkout. Ensure the proxy is running and the route
// exists, then run the dev command.
// ---------------------------------------------------------------------------

struct UpTarget {
    project: String,
    domain: String,
    port: u16,
    path: String,
}

pub fn run(open: bool) {
    if let Err(e) = run_inner(&cwd_string(), open) {
        eprintln!("\n  {} {}\n", red("Error:"), e);
    }
}

fn run_inner(cwd: &str, open: bool) -> Result<()> {
    let workspaces = config::load_workspaces()?;

    let target = if let Some(w) = locate_workspace(&workspaces, cwd, None) {
        UpTarget {
            project: w.project.clone(),
            domain: w.proxy_domain.clone(),
            port: w.port,
            path: w.path.clone(),
        }
    } else {
        let ship_config = config::load_config()?;
        let entry = ship_config
            .projects
            .iter()
            .find(|(_, p)| cwd == p.path || cwd.starts_with(&format!("{}/", p.path)));
        let Some((alias, root)) = entry else {
            println!(
                "  {} Not inside a workspace or registered project. Use 'ship init' or 'ship create' first.",
                red("✗")
            );
            return Ok(());
        };
        let alias = alias.clone();
        let domain = root
            .domain
            .clone()
            .unwrap_or_else(|| format!("{alias}.localhost"));
        let port = match root.port {
            Some(p) => p,
            None => {
                // Project registered before root routes existed — allocate and persist.
                let p = proxy::next_port()?;
                let mut updated = root.clone();
                updated.domain = Some(domain.clone());
                updated.port = Some(p);
                config::add_project(&alias, updated)?;
                p
            }
        };
        UpTarget {
            project: alias,
            domain,
            port,
            path: root.path.clone(),
        }
    };

    let project_config = config::get_project(&target.project)?;

    // Ensure proxy is running.
    if !proxy::is_running() {
        proxy::start()?;
        println!("  {} Proxy started.", green("●"));
    }

    // Ensure route exists.
    let _ = proxy::add_route(&target.domain, target.port);

    println!(
        "  {} {} {} localhost:{}",
        green("●"),
        bold(&target.domain),
        dim("→"),
        blue(target.port)
    );

    // Resolve dev command sequence.
    let dev = project_config.commands.dev.clone();
    let Some((last, rest)) = dev.split_last() else {
        println!("  {}", dim("No dev command configured. Proxy route is active."));
        return Ok(());
    };
    let resolved_last = last.replace("{port}", &target.port.to_string());

    // Open browser after a short delay, in the background.
    if open {
        let url = format!("https://{}", target.domain);
        std::thread::spawn(move || {
            std::thread::sleep(std::time::Duration::from_secs(2));
            let _ = std::process::Command::new("open").arg(&url).output();
        });
    }

    // Run any preparatory dev commands to completion, in order.
    for cmd in rest {
        let resolved = cmd.replace("{port}", &target.port.to_string());
        println!("  {}", dim(format!("Running: {resolved}")));
        shell::exec_in_dir(&target.path, &resolved, &[])?;
    }

    println!("  {}", dim(format!("Running: {resolved_last}")));
    println!();

    // Run the final dev command (blocks until it exits — the dev server).
    shell::exec_in_dir(&target.path, &resolved_last, &[])?;
    Ok(())
}
